python/semantic_kernel/connectors/memory/azure_ai_search/azure_ai_search_memory_record_service.py

```diff
@@ -0,0 +1,432 @@
+
+        Args:
+            collection_name (str): The name of the collection to check.
+
+        Returns:
+            bool: True if the collection exists; otherwise, False.
+        """
+        try:
+            collection_result = await self._search_index_client.get_index(name=collection_name.lower())
+
+            return bool(collection_result)
+        except ResourceNotFoundError:
+            return False
+
+    # endregion
+
+    @overload
+    async def upsert(self, collection_name: str, record: DataModelT) -> str:
+        """Upsert a record.
+
+        Args:
+            collection_name (str): The name of the collection to upsert the record into.
+            record (MemoryRecord): The record to upsert.
+
+        Returns:
+            str: The unique record id of the record.
+        """
+        result = await self.upsert_batch(collection_name, [record])
+        if result:
+            return result[0]
+        return None
+
+    @overload
+    async def upsert_batch(self, collection_name: str, records: list[DataModelT], **kwargs) -> list[str]:
+        """Upsert a batch of records.
+
+        Args:
+            collection_name (str): The name of the collection to upsert the records into.
+            records (List[MemoryRecord]): The records to upsert.
+
+        Returns:
+            List[str]: The unique database keys of the records.
+        """
+        # Initialize search client here
+        # Look up Search client class to see if exists or create
+        search_client = self._search_index_client.get_search_client(collection_name.lower())
+
+        search_records = []
+        search_ids = []
+
+        for record in records:
+            # Note:
+            # * Document id     = user provided value
+            # * MemoryRecord.id = base64(Document id)
+            if not record._id:
+                record._id = str(uuid.uuid4())
+
+            search_record = memory_record_to_search_record(record)
+            search_records.append(search_record)
+            search_ids.append(record._id)
+
+        result = await search_client.upload_documents(documents=search_records)
+        await search_client.close()
+
+        if result[0].succeeded:
+            return search_ids
+        return None
+
+    @overload
+    async def get(self, collection_name: str, key: str, with_embedding: bool = False, **kwargs) -> DataModelT:
+        """Gets a record.
+
+        Args:
+            collection_name (str): The name of the collection to get the record from.
+            key (str): The unique database key of the record.
+            with_embedding (bool): Whether to include the embedding in the result. (default: {False})
+
+        Returns:
+            MemoryRecord: The record.
+        """
+        # Look up Search client class to see if exists or create
+        search_client = self._search_index_client.get_search_client(collection_name.lower())
+
+        try:
+            search_result = await search_client.get_document(
```

Comment:
In C# I've adopted a pattern whereby there is a default mapper (in this case whatever Azure AI Search does to convert from the serialized stream into a model) but the user can also optionally inject their own mapper.  This is useful for cases where the data model that the user wants cannot be converted to automatically using the built in mapper from the storage SDK or whatever code we may have on top of that.  For azure ai search using the custom mapper, I'm instructing the azure ai search SDK to convert to json objects, which is an object model representing the raw json.  I'm then passing that into the mapper.  Not sure if the python azure ai search sdk supports doing something like that too though.  Let me know if it doesn't anyway.  I thought this is a nice approach in C# anyway, and I'm using it for the redis implementation too,